### PR TITLE
METRON-1448: Update SolrWriter to conform to new collection strategy

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
Currently the SolrWriter presumes a single collection to be written to.  The new collection strategy for Solr implies a collection per sensor.  Also, there are a few rough edges in the writer which could stand smoothing:
* By default, we use solr's implicit commit mechanism, rather than committing at the batch granularity.  This may result in lost data on worker failure.
* We do not use the the batch add api, but rather message-by-message add

# Test Plan for Solr Writer

The intent of this PR is to get the SolrWriter to the point where solr
can be used as a random access index.  There are some manual steps
required at this time due to this feature not being fully baked.  I will
ensure that those are obvious when I am stating them.

As usual, the assumption are:
* An environment variable `METRON_HOME` (set to `/usr/metron/0.4.3`)
* An environment variable `ZOOKEEPER` (set to `node1:2181`)

## Preliminaries (optional)
* Kill the profiler and batch indexing via:
```
storm kill batch_indexing && storm kill profiler
```

## Solr Setup 
Note that in the future this will be automated either via ansible or the
management pack in Ambari.  Unless otherwise stated, run the commands as
`root` (at least in full-dev, perhaps `metron` in a real cluster).

* Kill the `random_access_indexing` topology:
```
storm kill random_access_indexing
```
* Install the `metron-solr` RPM via `yum install metron-solr`
* Run the solr installation script via `$METRON_HOME/bin/install_solr.sh`
* Install the schemas, as the `solr` user (e.g. `sudo su - solr`)
```
/var/solr/solr-6.6.2/bin/solr create -c error -d /usr/metron/0.4.3/config/schema/error
/var/solr/solr-6.6.2/bin/solr create -c bro -d /usr/metron/0.4.3/config/schema/bro
/var/solr/solr-6.6.2/bin/solr create -c snort -d /usr/metron/0.4.3/config/schema/snort
```
* Fix a bug that will be corrected in the PR that ensures solr support
  in the management pack by editing
`$METRON_HOME/config/solr.properties`.
  * Change `topology.auto-credentials=['']` to `topology.auto-credentials=[]`
* It's unclear whether or not this is strictly required (probably not),
  but I adjusted the `group.id` for the `random_access_indexing` topology by editing `$METRON_HOME/flux/indexing/random_access/remote.yaml`
  * On line 67, change the `group.id` to `indexing-solr-ra`
  * This is required so the kafka offsets are reset and the solr
    topology pulls from the back of the queue rather than where the ES indexing topology left off.
* Edit `$METRON_HOME/config/zookeeper/global.json` and add the following
  property: `"solr.zookeeper" : "localhost:9983"` to the global config
* Push the config to zookeeper via `$METRON_HOME/bin/zk_load_util.sh -m PUSH -i $METRON_HOME/config/zookeeper -z $ZOOKEEPER`
* Start the solr indexing topology via `$METRON_HOME/bin/start_solr_topology.sh`

## Validation

You can go to `http://node1:8983/solr/#/bro/query` and
`http://node1:8983/solr/#/snort/query` and hit the `Execute Query`
button to run queries against solr to ensure data shows up in the
`snort` and `bro` collections

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/writer/SolrWriter.java [hunk 1]
```diff
@@ -33,17 +39,19 @@
 
 import java.io.IOException;
 import java.io.Serializable;
-import java.util.List;
-import java.util.Map;
+import java.lang.invoke.MethodHandles;
+import java.util.*;
```
Comment (JonZeolla) [inline metron-platform/metron-solr/src/main/java/org/apache/metron/solr/writer/SolrWriter.java]: Style police checking in.  No \* please.

Comment (cestella) [reply metron-platform/metron-solr/src/main/java/org/apache/metron/solr/writer/SolrWriter.java]: You got it :)

**Comment by ottobackwards** [comment]
Should we explore auto-commit?
Also, maybe we should rename all this stuff to SolrCloud, since Solr and SolrCloud *are* different in some use of terms and concepts?

**Comment by cestella** [comment]
What do you mean "auto-commit"?  By default it will commit at the batch level as the other writers do.  You can turn that behavior off and it'll commit when the Solr client decides at the cost of possibly losing data.

Regarding solr vs solrcloud, to my understanding solr is the product and solrcloud is a way of running Solr.  The other way is solr standalone.  Originally we named things after the product, which I think is correct, and have made explicit that we support Solr running in Solr Cloud mode in the docs.

**Comment by ottobackwards** [comment]
Sorry, maybe I misunderstood what I have read here: 
https://lucidworks.com/2013/08/23/understanding-transaction-logs-softcommit-and-commit-in-sorlcloud/

I thought that you could set to not have to manually commit.

**Comment by cestella** [comment]
You can set it to not manually commit (set `solr.commitPerBatch` to `false` and no committing happens), but you're risking losing data if a worker dies.  Honestly, want a durable commit with a fsync before you ack the tuples in a batch, otherwise you're courting data loss.  This is the same strategy we do for ES and HDFS (though commit there is a fsync).  That being said, I'm sensitive to performance issues around that that people may have, so I let people turn it off with a strong warning in the docs (also this was legacy behavior in the SolrWriter).

**Comment by cestella** [comment]
Actually, I can make an option to let you do a soft commit instead of a hard commit if that'll be an extra layer of configuration.

**Comment by ottobackwards** [comment]
See, we explored it.  Thanks!!

**Comment by cestella** [comment]
LOL, nah, I was just being crotchety.  Let me see what I can do and you tell me what you think :)

**Comment by cestella** [comment]
Ok, I updated the configs to expose all the settings that I could find for committing and documented them.  By default we operate safely, though.  Though now you can use soft commits rather than durable commits if you so desire and wish to live dangerously.

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/writer/SolrWriter.java (hunk 1 above)
Comment (cestella) [reply metron-platform/metron-solr/src/main/java/org/apache/metron/solr/writer/SolrWriter.java]: UGh, it slipped back in again. Ok, removing it.

**Comment by ottobackwards** [comment]
Cool, I'll run the test plan....

**Comment by cestella** [comment]
Ok, I wrote up a detailed testing plan to get this running in full-dev.  I want to add a couple of unit tests around the new properties that I added, but at that point, this is feature complete and ready to review.

**Comment by ottobackwards** [comment]
Why is it "solr.zookeeper" : "localhost:9983"  and not solr.ip?

**Comment by ottobackwards** [comment]
don't we have to push the new config to zookeeper?

**Comment by ottobackwards** [comment]
Everything worked, just wondering about the questions above before I check off.

**Comment by merrimanr** [comment]
+1 from me pending approval from others.  Tested this in full dev and worked fine.

**Comment by ottobackwards** [comment]
@merrimanr did you push the configs to zookeeper?

**Comment by merrimanr** [comment]
Yes I pushed the change to zookeeper with Swagger.  Probably just a minor detail he missed in his instructions.  

You can use either "localhost:9983" or "node1:9983".  That address is the embedded zookeeper that Solr starts up in this case but in general it is whatever zookeeper you configure Solr to use.

**Comment by ottobackwards** [comment]
+1

**Comment by cestella** [comment]
I updated the instructions and the `solr.zookeeper` business is legacy naming, but I think it does fit since it's specific to zookeeper rather than connecting to ES master.  I dunno, that's just what I told myself ;)

**Pull request closed without merging**